builder/virtualbox/common/step_attach_isos.go

```diff
@@ -0,0 +1,158 @@
+		// Get the guest additions path since we're doing it
+		guestAdditionsPath := state.Get("guest_additions_path").(string)
+		diskMountMap["guest_additions"] = guestAdditionsPath
+	}
+
+	if len(diskMountMap) == 0 {
+		ui.Message("No ISOs to mount; continuing...")
+		return multistep.ActionContinue
+	}
+
+	driver := state.Get("driver").(Driver)
+	vmName := state.Get("vmName").(string)
+
+	for diskCategory, isoPath := range diskMountMap {
+		// If it's a symlink, resolve it to its target.
+		resolvedIsoPath, err := filepath.EvalSymlinks(isoPath)
+		if err != nil {
+			err := fmt.Errorf("Error resolving symlink for ISO: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		isoPath = resolvedIsoPath
+
+		// We have three different potential isos we can attach, so let's
+		// assign each one its own spot so they don't conflict.
+		var controllerName, device, port string
+		switch diskCategory {
+		case "boot_iso":
+			// figure out controller path
+			controllerName = "IDE Controller"
+			port = "0"
+			device = "1"
+			if s.ISOInterface == "sata" {
+				controllerName = "SATA Controller"
+				port = "1"
+				device = "0"
+			}
+			ui.Message("Mounting boot ISO...")
+		case "guest_additions":
+			controllerName = "IDE Controller"
+			port = "1"
+			device = "0"
+			if s.GuestAdditionsInterface == "sata" {
+				controllerName = "SATA Controller"
+				port = "2"
+				device = "0"
+			}
+			ui.Message("Mounting guest additions ISO...")
+		case "cd_files":
+			controllerName = "IDE Controller"
+			port = "1"
+			device = "1"
+			if s.ISOInterface == "sata" {
+				controllerName = "SATA Controller"
+				port = "3"
+				device = "0"
+			}
+			ui.Message("Mounting cd_files ISO...")
+		}
+
+		// Attach the disk to the controller
+		command := []string{
+			"storageattach", vmName,
+			"--storagectl", controllerName,
+			"--port", port,
+			"--device", device,
+			"--type", "dvddrive",
+			"--medium", isoPath,
+		}
+		if err := driver.VBoxManage(command...); err != nil {
+			err := fmt.Errorf("Error attaching ISO: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+
+		// Track the disks we've mounted so we can remove them without having
+		// to re-derive what was mounted where
+		unmountCommand := []string{
+			"storageattach", vmName,
+			"--storagectl", controllerName,
+			"--port", port,
+			"--device", device,
+			"--type", "dvddrive",
+			"--medium", "none",
+		}
+
+		s.diskUnmountCommands[diskCategory] = unmountCommand
+	}
+
+	state.Put("disk_unmount_commands", s.diskUnmountCommands)
+
+	return multistep.ActionContinue
+}
+
+func (s *StepAttachISOs) Cleanup(state multistep.StateBag) {
+	if len(s.diskUnmountCommands) == 0 {
+		return
+	}
+
+	driver := state.Get("driver").(Driver)
+
+	for _, command := range s.diskUnmountCommands {
+		// Remove the ISO. Note that this will probably fail since
+		// stepRemoveDevices does this as well. No big deal.
```

Comment:
Nice! I was wondering this before reading the comment. 
If you wish, you could just remove the diskUnmountCommands from the state and use it here instead of the diskUnmountCommands variable. 
```
state.Remove("disk_unmount_commands") // in step_remove_devices.go after removal
```

and the if statement that checks the length of diskUnmountCommands would check if "disk_unmount_commands" exists in the state.

Anyways, the way that it is right now I think is great! Don't mind changing if you're also ok with it 👍

Reply:
Yeah, I agree -- I'll add a note to the statebag so we don't waste operations